### PR TITLE
Part of fix/code-exchange: Add code-exchange flow to JolliAuthService

### PR DESCRIPTION
**Pull request by jordangreen-jolli**
## E2E Test (4)
<details>
<summary><strong>1. Sign-in completes successfully after browser approval</strong></summary>
<br>
<blockquote>

**Preconditions:** Have the IntelliJ plugin installed and be signed out. Ensure you have a working internet connection and valid account credentials.

**Steps:**
1. Open IntelliJ and click the JolliMemory plugin icon or go to Settings and find the JolliMemory sign-in option.
2. Click "Sign In" to start the login flow — your browser should open automatically to the Jolli login page.
3. In the browser, approve access when prompted (log in with your account if needed and click "Allow" or "Authorize").
4. Switch back to IntelliJ and wait a few seconds.
5. Check that the plugin no longer shows a loading or "waiting" state.
6. Verify that your account name or a "signed in" indicator appears in the plugin panel.

**Expected Results:**
- The browser shows a success page after you approve access.
- IntelliJ does not show a timeout error or "couldn't reach Jolli" message.
- The plugin panel shows you as signed in, with no error dialogs appearing.

</blockquote>
</details>
<details>
<summary><strong>2. Expired or already-used sign-in link shows a clear error</strong></summary>
<br>
<blockquote>

**Preconditions:** Have the IntelliJ plugin installed and be signed out.

**Steps:**
1. Start the sign-in flow in IntelliJ by clicking "Sign In" — your browser opens to the Jolli login page.
2. Do not complete the login in the browser. Instead, wait several minutes (or close and restart the sign-in flow to generate a fresh link, making the first one stale).
3. Go back to the original browser tab with the old login link and try to approve access there.
4. Switch back to IntelliJ and observe what message appears.

**Expected Results:**
- The plugin displays an error message indicating the sign-in code has expired or was already used.
- The message prompts you to try signing in again.
- No crash or blank/frozen state occurs — the plugin remains usable so you can start a new sign-in attempt.

</blockquote>
</details>
<details>
<summary><strong>3. Cancelled sign-in shows a friendly message</strong></summary>
<br>
<blockquote>

**Preconditions:** Have the IntelliJ plugin installed and be signed out.

**Steps:**
1. Click "Sign In" in the JolliMemory plugin panel — your browser opens to the Jolli login page.
2. On the authorization page in the browser, click "Deny", "Cancel", or close the approval dialog without approving.
3. Switch back to IntelliJ and observe the message shown in the plugin.

**Expected Results:**
- The plugin displays a message along the lines of "Sign-in was cancelled. You can try again from Settings."
- No generic or cryptic error code is shown to you.
- The plugin remains usable and you can attempt to sign in again.

</blockquote>
</details>
<details>
<summary><strong>4. Sign-in error messages are human-readable for all failure types</strong></summary>
<br>
<blockquote>

**Preconditions:** Have access to a test or staging Jolli environment where specific error conditions (such as unverified email, invalid provider, or session expiry) can be triggered during login.

**Steps:**
1. Attempt to sign in using an account that has no verified email address on file, and observe the error message shown in IntelliJ.
2. Attempt to sign in using an unsupported or misconfigured authentication provider, and observe the error message.
3. Trigger a session-expiry scenario (for example, by letting a login session time out on the server side) and observe the error message.
4. For each case, check that the message shown is plain English with a clear explanation.

**Expected Results:**
- The "no verified email" error shows a message mentioning verified email addresses, not a raw code like "no_verified_emails".
- The "invalid provider" error shows a message mentioning the authentication provider.
- The "session expired" error shows a message mentioning the session and prompts you to try again.
- In all cases, the message is readable by a non-technical user with no error codes or jargon visible.

</blockquote>
</details>

---

## Topics (3)
<details>
<summary><strong>01 · Add secure code-exchange login flow to IntelliJ plugin auth</strong></summary>
<br>
<blockquote>

**⚡ Why This Change**

The IntelliJ plugin's sign-in flow was broken: after the user approved access in the browser, the plugin would hang and eventually show a timeout error saying it couldn't reach Jolli to complete sign-in. The callback was arriving correctly, but the second step of exchanging the authorization code for credentials was failing.

**💡 Decisions Behind the Code**

- **Same URL for login and exchange**: The plugin uses the same `jolliUrl` for both opening the browser login page and POSTing the code exchange, mirroring the CLI and VSCode extension. In production, the frontend and API share a host, so this works without a separate config value. Local environments where the frontend and backend are on different ports require pointing `JOLLI_URL` at the backend or using a deployed URL.
- **Legacy token fallback**: Rather than dropping support for pre-JOLLI-1270 servers, the callback handler falls back to the old `?token=` path with a warning log, keeping the plugin compatible with older server deployments.
- **Blocking HTTP call accepted**: The exchange POST uses a synchronous `HttpClient.send()` on the callback thread, matching the CLI's pattern. This is acceptable because only one callback is expected per login flow.

**✅ What Was Implemented**

- Added `exchangeCode()` to `JolliAuthService.kt`: POSTs the single-use code to `/api/auth/cli-exchange` with a 20-second timeout, parses the JSON response, and returns a `LoginResult`; includes tenant-slug header support for path-based URLs.
- Rewrote the callback handler to prefer `?code=` (JOLLI-1270 secure path) and fall back to `?token=` (legacy) with a warning log, then unified the credential-saving logic into a single branch.
- Extended `getErrorMessage()` with eight new error codes introduced by the code-exchange flow (`oauth_failed`, `session_missing`, `no_verified_emails`, `user_denied`, `invalid_callback`, `failed_to_get_token`, `auth_fetch_failed`, `invalid_provider`).

**📁 FILES**
- `intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt`
- `intellij/src/test/kotlin/ai/jolli/jollimemory/services/JolliAuthServiceTest.kt`

</blockquote>
</details>
<details>
<summary><strong>02 · Add unit tests for code-exchange flow and new error codes</strong></summary>
<br>
<blockquote>

**⚡ Why This Change**

The new code-exchange logic needed test coverage to verify correct behavior across success, failure, and edge-case scenarios.

**💡 Decisions Behind the Code**

A real embedded HTTP server was used for `exchangeCode` tests rather than mocking the HTTP client, so the full request/response cycle — including header inspection and status code handling — is exercised without faking internal implementation details.

**✅ What Was Implemented**

- Added a `startFakeExchangeServer()` helper that spins up a local HTTP server on a random port to simulate the exchange endpoint.
- Added eight `exchangeCode` tests covering: success with all fields, success with optional fields absent, 404 (expired code), non-OK status, missing token in response, malformed JSON, network error, and tenant-slug header presence/absence.
- Added `getErrorMessage` tests asserting human-readable strings for all eight new JOLLI-1270 error codes.

**📁 FILES**
- `intellij/src/test/kotlin/ai/jolli/jollimemory/services/JolliAuthServiceTest.kt`

</blockquote>
</details>
<details>
<summary><strong>03 · Remove unused import and debug log line from auth service</strong></summary>
<br>
<blockquote>

**⚡ Why This Change**

During a code review, an unused import and a temporary debug log line were identified in the auth service and flagged for cleanup before merging.

**💡 Decisions Behind the Code**

Both removals were straightforward cleanup with no alternatives considered.

**✅ What Was Implemented**

- Removed the unused `JolliMemoryConfig` import from `JolliAuthService.kt`.
- Removed the debug log line that printed the code prefix during the exchange flow.

**📁 FILES**
- `intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt`

</blockquote>
</details>

---

*Generated by JolliMemory · May 6, 2026 at 6:21 PM*

### CODE REVIEW

**Comment by flyerli2025** [comment]
Findings:

P1 intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:110-116
The new state check runs before the callback is classified as ?error=, ?code=, or legacy ?token=. That breaks two flows this PR says it preserves: error callbacks like user_denied and old-server token callbacks, both of which are handled without state in the CLI path. In practice those requests will now be rejected as “Invalid login state” before getErrorMessage() or the legacy fallback can run, so canceled sign-in and pre-JOLLI-1270 compatibility regress.

P2 intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:231-233
exchangeCode() derives tenantSlug with parsed.path.trimStart('/'), which keeps trailing slashes and extra path segments. For a common local URL like https://jolli-local.me/dev/, this sends x-tenant-slug: dev/ instead of dev. The shared parsers used elsewhere only take the first normalized path segment, so this Kotlin port can still fail against path-based local/staging tenants even though the added test only covers /dev without the trailing slash.

**Review by dougschroeder — COMMENTED** [review]
This cleanly mirrors the CLI's wire shape and adds thorough tests for the happy and unhappy paths. Two findings worth attention before merge: the new credential-bearing POST should re-validate `jolliUrl` against the same allowlist the CLI uses, and the returned `jolliApiKey` should be passed through `parseJolliApiKey` (or equivalent) before persistence — both gaps that would otherwise let a tampered `JOLLI_URL` redirect or inject credentials. Everything else is parity polish.

**General finding (not tied to a single line):** Extract a shared Kotlin auth-utils file. `parseJolliApiKey` lives in `JolliApiClient.kt`; the proposed `assertJolliOriginAllowed` and `validateJolliApiKey` belong with it. Consolidate into `intellij/src/main/kotlin/ai/jolli/jollimemory/auth/JolliAuthUtils.kt` so the IntelliJ services share one canonical port of `cli/src/core/JolliApiUtils.ts`. Also consolidate `parseBaseUrl` (currently in `JolliApiClient.kt:320` and inlined again in `exchangeCode`).

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 1]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
         }
     }
 
+    /**
+     * Exchanges a single-use authorization code for credentials via POST to
+     * `/api/auth/cli-exchange`. Mirrors the CLI's `exchangeCliCode()`.
+     */
+    internal fun exchangeCode(jolliUrl: String, code: String): LoginResult {
+        val parsed = URI.create(jolliUrl)
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:238]: **Show-stopper:** `exchangeCode` issues a credential-bearing POST without re-validating `jolliUrl` against the Jolli origin allowlist or enforcing HTTPS.

**Why it matters:** The CLI counterpart calls `assertJolliOriginAllowed(jolliUrl)` at the top of `exchangeCliCode` (`cli/src/auth/CliExchange.ts:49`), and CLAUDE.md says the three implementations (CLI, VSCode, IntelliJ) must stay in lockstep. The IntelliJ port reads `JOLLI_URL` from env/system property at every call with no validation, so a tampered/stale value sends the user's browser **and** the single-use code to an arbitrary host. The allowlist used by the CLI is `jolli.ai`, `jolli.dev`, `jolli.cloud`, `jolli-local.me`, HTTPS-only, with suffix-boundary checks (`host == h || host.endsWith(".$h")`).

**Suggested fix:** Port `assertJolliOriginAllowed` to Kotlin and call it at the top of `exchangeCode` (and ideally at `JolliUrlConfig.getJolliUrl()` save-time too). Existing tests at lines 142-265 use `http://127.0.0.1:port` and will need a `dangerouslyAllowedForTests` overload.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 2]
```diff
@@ -117,30 +123,50 @@ object JolliAuthService {
...
 
-                            val html = successHtml()
-                            sendHtml(exchange, html, 200)
-                            notifyAuthListeners()
-                            onSuccess(LoginResult(token, space, jolliApiKey))
+                        val globalDir = SessionTracker.getGlobalConfigDir()
+                        val existing = SessionTracker.loadConfigFromDir(globalDir)
+                        SessionTracker.saveConfigToDir(existing.copy(
+                            authToken = result.token,
+                            jolliApiKey = if (!result.jolliApiKey.isNullOrBlank()) result.jolliApiKey else existing.jolliApiKey,
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:165]: **Show-stopper:** The `jolliApiKey` returned from `exchangeCode` is persisted to the global config without validation.

**Why it matters:** The CLI runs `parseJolliApiKey` to decode the key and check the embedded `u` value against the origin allowlist before save (`cli/src/core/JolliApiUtils.ts`). `JolliApiClient.kt:127-150` already has a Kotlin port of `parseJolliApiKey`, but it's never invoked here. Without this guard, a compromised exchange endpoint can plant a key whose embedded URL points anywhere — turning the new exchange path into a credential-injection vector.

**Suggested fix:** Extend `parseJolliApiKey` (or add a `validateJolliApiKey` companion) so it throws on null-decode or off-allowlist `u`, then call it before `SessionTracker.saveConfigToDir(...)`.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 3]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
         }
     }
 
+    /**
+     * Exchanges a single-use authorization code for credentials via POST to
+     * `/api/auth/cli-exchange`. Mirrors the CLI's `exchangeCliCode()`.
+     */
+    internal fun exchangeCode(jolliUrl: String, code: String): LoginResult {
+        val parsed = URI.create(jolliUrl)
+        val origin = "${parsed.scheme}://${parsed.authority}"
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt]: **Medium Priority:** `${parsed.scheme}://${parsed.authority}` includes user-info (`user:pass@host:port`) when the URL has a userinfo component.

**Why it matters:** `URI.authority` returns `user:pass@host:port` when userinfo is present, so `JOLLI_URL=https://x:y@host` would propagate basic-auth credentials into the exchange URL. The CLI uses `new URL(...).origin` which is host+port+scheme only, so the two implementations diverge here.

**Suggested fix:** Build origin manually:

```kotlin
val origin = "${parsed.scheme}://${parsed.host}${if (parsed.port >= 0) ":${parsed.port}" else ""}"
```

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 4]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
...
+        val tenantSlug = parsed.path.split('/').firstOrNull { it.isNotEmpty() }
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt]: **Medium Priority:** `parsed.path.split('/')` will NPE for URIs whose path is null (opaque URIs, malformed input).

**Why it matters:** `URI.path` is a Java platform type (`String!`); calling `.split` on null throws NPE without a clear stack frame. `JolliApiClient.kt:322` already handles this defensively with `(uri.path ?: "")`.

**Suggested fix:** `(parsed.path ?: "").split('/').firstOrNull { it.isNotEmpty() }`. Better still, extract a shared `parseBaseUrl` so this and `JolliApiClient` stay in lockstep.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 5]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
...
+
+        if (tenantSlug != null) {
+            requestBuilder.header("x-tenant-slug", tenantSlug)
+        }
+
+        val response: HttpResponse<String>
+        try {
+            response = HttpClient.newHttpClient()
+                .send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString())
+        } catch (e: Exception) {
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:260]: **Medium Priority:** `HttpTimeoutException` (raised by the `.timeout(Duration.ofSeconds(20))` at line 241) is caught generically and emitted as `"Couldn't reach Jolli to complete sign-in: …"`.

**Why it matters:** The CLI distinguishes the timeout case from generic connectivity errors (`cli/src/auth/CliExchange.ts:67-73`), surfacing `"Sign-in timed out after 20s waiting for Jolli. Please try again."` separately. This matters for UX — slow-network vs wrong-URL produce very different user actions.

**Suggested fix:** Catch `java.net.http.HttpTimeoutException` first and emit a timeout-specific message that mirrors the CLI's wording.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 6]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
...
+        if (response.statusCode() == 404) {
+            throw RuntimeException("Sign-in code expired or already used. Please try signing in again.")
+        }
+        if (response.statusCode() !in 200..299) {
+            throw RuntimeException("Sign-in failed (HTTP ${response.statusCode()}).")
+        }
+
+        val payload: JsonObject
+        try {
+            payload = Gson().fromJson(response.body(), JsonObject::class.java)
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:273]: **Suggestion:** The ad-hoc `Gson().fromJson(…, JsonObject::class.java)` plus three `payload.get(…)?.takeIf { it.isJsonPrimitive }?.asString` calls would read more cleanly as a typed data class:

```kotlin
data class CliExchangeResponse(
    val token: String? = null,
    val jolliApiKey: String? = null,
    val space: String? = null,
)
```

The missing-token guard then becomes a simple non-null check on `response.token`.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 7]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
...
+
+        val token = payload.get("token")?.takeIf { it.isJsonPrimitive }?.asString
+        if (token.isNullOrBlank()) {
+            throw RuntimeException("Sign-in failed: server response did not include a token.")
+        }
+
+        val jolliApiKey = payload.get("jolliApiKey")?.takeIf { it.isJsonPrimitive }?.asString
+        val space = payload.get("space")?.takeIf { it.isJsonPrimitive }?.asString
+
+        return LoginResult(token = token, space = space, jolliApiKey = jolliApiKey)
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:287]: **Suggestion:** The successful exchange path has no log breadcrumb. `Login.ts` in the CLI emits a one-line summary on success (covering only the *fact* of success, never the token). Useful for support diagnostics when users start hitting the new flow.

**Suggested fix:** Something like

```kotlin
log.info("cli-exchange completed (apiKey present: %s, space: %s)", result.jolliApiKey != null, result.space ?: "<none>")
```

before returning.

File: intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt [hunk 8]
```diff
@@ -198,11 +224,73 @@ object JolliAuthService {
...
+    }
+
     internal fun getErrorMessage(code: String): String = when (code) {
         "access_denied" -> "Access was denied. Please try again."
         "invalid_request" -> "Invalid login request. Please try again."
         "server_error" -> "Server error during login. Please try again later."
         "temporarily_unavailable" -> "Service temporarily unavailable. Please try again later."
+        "oauth_failed" -> "OAuth authentication failed. Please try again."
```
Comment (dougschroeder) [inline intellij/src/main/kotlin/ai/jolli/jollimemory/services/JolliAuthService.kt:295]: **Suggestion:** Group the new code-exchange error codes under a comment so future readers can see at a glance which codes correspond to which flow.

```kotlin
// Code-exchange error codes (returned via ?error= on the callback URL)
"oauth_failed" -> ...
```

File: intellij/src/test/kotlin/ai/jolli/jollimemory/services/JolliAuthServiceTest.kt [hunk 9]
```diff
@@ -108,6 +111,159 @@ class JolliAuthServiceTest {
...
+            exchange.sendResponseHeaders(statusCode, bytes.size.toLong())
+            exchange.responseBody.use { it.write(bytes) }
+        }
+        server.start()
+        return server
+    }
+
+    @Test
+    fun `exchangeCode returns token and optional fields on success`() {
+        val server = startFakeExchangeServer(200, """{"token":"jwt-abc","jolliApiKey":"sk-jol-123","space":"my-space"}""")
```
Comment (dougschroeder) [inline intellij/src/test/kotlin/ai/jolli/jollimemory/services/JolliAuthServiceTest.kt:153]: **Suggestion:** Significant `try { … } finally { server.stop(0) }` boilerplate across these tests. Collapse into a `withFakeExchangeServer(statusCode, body) { server -> … }` helper, or a small `FakeJolliExchangeServer` fixture, especially since two tests at lines 230-265 drop the helper entirely so they can capture headers.

**Review by dougschroeder — APPROVED** [review]
